python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,726 @@
+
+        f = open(USER_DATA_FILE_PATH, "r")
+        data = yaml.load(f, Loader=yaml.FullLoader)
+        for _, v in data.items():
+            for user in v:
+                if isinstance(user, dict):
+                    user["ssh_authorized_keys"] = [public_key]
+        f.close()
+
+        modified_userdata = yaml.dump(data, default_flow_style=False)
+
+        # The userdata needs to be prefixed with #cloud-config.
+        # Without it, the cloudinit spec would get applied but
+        # it wouldn't add the userdata on the VM.
+
+        modified_userdata = "#cloud-config\n" + modified_userdata
+        logger.info("Successfully modified the userdata file for vm {}".format(vm_id))
+
+        # Create cloud-init spec and apply
+        cloudinit_config = CloudinitConfiguration(
+            metadata=metadata, userdata=modified_userdata
+        )
+        cloud_config = CloudConfiguration(
+            cloudinit=cloudinit_config, type=CloudConfiguration.Type("CLOUDINIT")
+        )
+        config_spec = ConfigurationSpec(cloud_config=cloud_config)
+        global_dns_settings = GlobalDNSSettings()
+        adapter_mapping_list = []
+        customization_spec = CustomizationSpec(
+            configuration_spec=config_spec,
+            global_dns_settings=global_dns_settings,
+            interfaces=adapter_mapping_list,
+        )
+
+        # create customization specification by CustomizationSpecs service
+        specs_svc = self.vsphere_sdk_client.vcenter.guest.CustomizationSpecs
+        spec_name = str(uuid.uuid4())
+        spec_desc = (
+            "This is a customization specification which includes"
+            "raw cloud-init configuration data"
+        )
+        create_spec = specs_svc.CreateSpec(
+            name=spec_name, description=spec_desc, spec=customization_spec
+        )
+        specs_svc.create(spec=create_spec)
+
+        vmcust_svc = self.vsphere_sdk_client.vcenter.vm.guest.Customization
+        set_spec = vmcust_svc.SetSpec(name=spec_name, spec=None)
+        vmcust_svc.set(vm=vm_id, spec=set_spec)
+
+        logger.info("Successfully added cloudinit config for vm {}".format(vm_id))
+
+    # Example: If a tag called node-status:initializing is present on the VM.
+    # If we would like to add a new value called finished with the node-status key.
+    # We'll need to delete the older tag node-status:initializing first before creating
+    # node-status:finished
+    def remove_tag_from_vm(self, tag_key_to_remove, vm_id):
+        yn_id = DynamicID(type=Constants.TYPE_OF_RESOURCE, id=vm_id)
+
+        # List all the tags present on the VM.
+        for tag_id in self.vsphere_sdk_client.tagging.TagAssociation.list_attached_tags(
+            yn_id
+        ):
+            vsphere_vm_tag = self.vsphere_sdk_client.tagging.Tag.get(tag_id=tag_id).name
+            tag_key_value = vsphere_tag_to_kv_pair(vsphere_vm_tag)
+            tag_key = tag_key_value[0] if tag_key_value else None
```

Comment:
Does it make sense to use get_matched_tags here too, maybe with an empty filter list? Just to avoid rewriting code

Reply:
Here we need the tag_id, which is the object id in vSphere to delete the tag, that is what get_matched_tags cannot provide.

Thanks for the suggestion.